### PR TITLE
Faster F-domain waveforms

### PR DESCRIPTION
**Pull request by spxiwh**
I would now like to merge this branch that Alex and I have been working on. Basically it adds a new set of approximants ending in "_INTERP", which are generated as the usual waveform (ie. IMRPhenomD_INTERP is PhenomD) but use a delta_f corresponding to the actual waveform length, and then uses an IFFT + zero-padding + FFT process to upsample the waveform to the desired time length (delta_f). I will present this on today's call.

Using this we see significantly faster throughput for F-domain waveforms *especially* IMRPhenomD, which when we tested was quite a bit slower than the ROM in our search codes. Here is a benchmark for SEOBNRv2_ROM_DoubleSpin:

https://atlas1.atlas.aei.uni-hannover.de/~ahnitz/LSC/ahnitz/projects/profile/goutput3.png

and for SEOBNRv2_ROM_DoubleSpin_INTERP:

https://atlas1.atlas.aei.uni-hannover.de/~ahnitz/LSC/ahnitz/projects/profile/goutput4.png

Here is a faithsim between SEOBNRv2_ROM_DoubleSpin and SEOBNRv2_ROM_DoubleSpin_INTERP:

https://galahad.aei.mpg.de/~spxiwh/LVC/waveforms/inter_test/faithsim/

all matches are > 0.996 and most are > 0.999

the same for PhenomD:

https://galahad.aei.mpg.de/~spxiwh/LVC/waveforms/inter_test/faithsim_phenomD/

and all matches are > 0.997

This doesn't affect existing codes as this adds *new* approximants rather than changing old ones. But this would offer us some speed improvement, especially if using PhenomD, or if using numerous cores or maybe GPUs.

### CODE REVIEW

**Comment by SebastianKhan-LIGO** [comment]
Ian, which version of lalsuite are you using?
During the review of PhenomD there was significant optimisation done
to PhenomD which places it on par with the ROMs.

The commit for the reviewed PhenomD waveform is
db16d17013531cd10451c7d0c6906972ce731866

**Comment by titodalcanton** [comment]
Ian, I'm curious to see the variation of the (small) unfaithfulness over the param space. Would you mind redoing your scatter plots with the colorbar limits from min to max faithfulness?

File: setup.py [hunk 1]
```diff
@@ -52,7 +52,7 @@
 requires = ['lal.lal', 'lalsimulation.lalsimulation', 'glue', 'pylal']
 setup_requires = []
 install_requires =  setup_requires + ['Mako>=1.0.1',
-                      'argparse>=1.3.0',
+                      'argparse>=1.2.0',
```
Comment (titodalcanton) [inline setup.py]: I'm happy with this change, but it should probably be split off to a new PR.

**Comment by spxiwh** [comment]
Hi Seb, you are right I was using an older PhenomD in these tests. The newer version is considerably faster, well done! Here's a short benchmark with PhenomD (not the new INTERP version) using vulcan hardware:

https://galahad.aei.mpg.de/~spxiwh/LVC/aLIGO/profile/goutput256TEST3.png

Here I'm seeing roughly 2/3 of time filtering vs 1/3 of time waveform generating. This seems to be about a factor of 2 slower than the ROMs, but this is much improved over my previous run with the older PhenomD. *However* using the IMRPhenomD interp version the profile looks the same as the SEOBNRv2 version. So this is about a 50% speed improvement on vulcan comparing new PhenomD vs new PhenomD_INTERP. Alex reports that when using mkl the FFTs are faster and you see even more of a speed improvement.

So, in summary, the optimized version is faster (which will greatly benefit PE!) but I might still suggest using IMRPhenomD_INTERP or SEOBNRv2_ROM_DoubleSpin_INTERP in searches. The speedup will likely be much more when using multi-threaded options.

File: setup.py (hunk 1 above)
Comment (spxiwh) [inline setup.py]: @ahnitz Why is this needed? I think you added this.

**Comment by spxiwh** [comment]
Hi Tito,

Good point! I've updated the colorbars on the faithsim to show what you wanted. Basically "worst" matches are at high eta.

Cheers
Ian

**Comment by SebastianKhan-LIGO** [comment]
Hi Ian, Good to see! Thanks!

**Comment by ahnitz** [comment]
Ian, Seb,

You also can get an additional, nearly factor of 2x speedup for high mass systems where the frequency stop of the waveform is below the filtering nyquist. We are doing this from the SEOBNRv2_ROM_DoubleSpin, and we could do it for IMRPhenomD as well if it provide a function that told us the final frequency for a given set of parameters. This makes the cost of the first FFT much more trivial, and so the dominant cost is only the single full length FFT at the end.

File: setup.py (hunk 1 above)
Comment (ahnitz) [inline setup.py]: It isn't. Feel free to remove.

Comment (spxiwh) [inline setup.py]: This has been removed from the patch now.

**Comment by duncan-brown** [comment]
@spxiwh please can you make a CBC wiki page summarizing the key results?

**Comment by ahnitz** [comment]
closing as this version has conflicts. I've opened a new pull request that is built ontop of master.
